File: Raven.Abstractions/Replication/ReplicationDestination.cs
//-----------------------------------------------------------------------
// <copyright file="ReplicationDestination.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.ComponentModel;

using Raven.Abstractions.Cluster;

namespace Raven.Abstractions.Replication
{
    /// <summary>
    /// Data class for replication destination documents
    /// </summary>
    public class ReplicationDestination
    {
        /// <summary>
        /// The name of the connection string specified in the 
        /// server configuration file. 
        /// Override all other properties of the destination
        /// </summary>

        private string url;

        /// <summary>
        /// Gets or sets the URL of the replication destination
        /// </summary>
        /// <value>The URL.</value>
        public string Url
        {
            get { return url; }
            set 
            {
                url = value.EndsWith("/") ? value.Substring(0, value.Length - 1) : value;
            }
        }

        /// <summary>
        /// The replication server username to use
        /// </summary>
        public string Username { get; set; }
        
        /// <summary>
        /// The replication server password to use
        /// </summary>
        public string Password { get; set; }

        /// <summary>
        /// The replication server domain to use
        /// </summary>
        public string Domain { get; set; }

        /// <summary>
        /// The replication server api key to use
        /// </summary>
        public string ApiKey { get; set; }

        /// <summary>
        /// The database to use
        /// </summary>
        public string Database { get; set; }

        /// <summary>
        /// How should the replication bundle behave with respect to replicated documents.
        /// If a document was replicated to us from another node, should we replicate that to
        /// this destination, or should we replicate only documents that were locally modified.
        /// </summary>
        public TransitiveReplicationOptions TransitiveReplicationBehavior { get; set; }

        /// <summary>
        /// Gets or sets a flag that controls whether index is replicated to the node or not
        /// </summary>
        public bool SkipIndexReplication { get; set; }

        /// <summary>
        /// Gets or sets if the replication will ignore this destination in the client
        /// </summary>
        public bool IgnoredClient { get; set; }

        /// <summary>
        /// Gets or sets if replication to this destination is disabled in both client and server.
        /// </summary>
        public bool Disabled { get; set; }

        public string AuthenticationScheme { get; set; }

        /// <summary>
        /// Gets or sets the Client URL of the replication destination
        /// </summary>
        public string ClientVisibleUrl { get; set; }

        /// <summary>
        /// If not null then only docs from specified collections are replicated and transformed / filtered according to an optional script.
        /// </summary>
        public Dictionary<string, string> SpecifiedCollections { get; set; }

        /// <summary>
        /// Gets or sets if attachments should be replicated when using ETL
        /// </summary>
        public bool ReplicateAttachmentsInEtl { get; set; } 

        public string Humane
        {
            get
            {
                if (string.IsNullOrEmpty(url))
                    return null;
                return url + " " + Database;
            }
        }

        public bool CanBeFailover()
        {
            return IgnoredClient == false && Disabled == false && (SpecifiedCollections == null || SpecifiedCollections.Count == 0);
        }

        protected bool Equals(ReplicationDestination other)
        {
            return IsEqualTo(other);
        }

        public bool IsEqualTo(ReplicationDestination other)
        {
            return string.Equals(Username, other.Username) && string.Equals(Password, other.Password) &&
                   string.Equals(Domain, other.Domain) && string.Equals(ApiKey, other.ApiKey) &&
                   string.Equals(Database, other.Database, StringComparison.OrdinalIgnoreCase) &&
                   TransitiveReplicationBehavior == other.TransitiveReplicationBehavior &&				   
                   IgnoredClient.Equals(other.IgnoredClient) && Disabled.Equals(other.Disabled) &&
                   ((string.Equals(Url, other.Url, StringComparison.OrdinalIgnoreCase) && string.IsNullOrWhiteSpace(ClientVisibleUrl)) ||
                   (!string.IsNullOrWhiteSpace(ClientVisibleUrl) && string.Equals(ClientVisibleUrl, other.ClientVisibleUrl, StringComparison.OrdinalIgnoreCase))) &&
                   ReplicateAttachmentsInEtl.Equals(other.ReplicateAttachmentsInEtl) && 
                   Extensions.DictionaryExtensions.ContentEquals(SpecifiedCollections, other.SpecifiedCollections);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((ReplicationDestination)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = (Username != null ? Username.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (Password != null ? Password.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (Domain != null ? Domain.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (ApiKey != null ? ApiKey.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (Database != null ? Database.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (int)TransitiveReplicationBehavior;
                hashCode = (hashCode * 397) ^ IgnoredClient.GetHashCode();
                hashCode = (hashCode * 397) ^ Disabled.GetHashCode();
                hashCode = (hashCode * 397) ^ (ClientVisibleUrl != null ? ClientVisibleUrl.GetHashCode() : 0);
                return hashCode;
            }
        }

        public class ReplicationDestinationWithConfigurationOrigin : ReplicationDestination
        {
            public bool HasGlobal { get; set; }

            public bool HasLocal { get; set; }
        }

        public class ReplicationDestinationWithClusterInformation : ReplicationDestination
        {
            public ClusterInformation ClusterInformation { get; set; }

            public static ReplicationDestinationWithClusterInformation Create(ReplicationDestinationWithConfigurationOrigin source, bool isInCluster, bool isLeader)
            {
                return new ReplicationDestinationWithClusterInformation
                       {
                           ApiKey = source.ApiKey,
                           ClientVisibleUrl = source.ClientVisibleUrl,
                           Database = source.Database,
                           Disabled = source.Disabled,
                           Domain = source.Domain,
                           IgnoredClient = source.IgnoredClient,
                           ClusterInformation = new ClusterInformation(isInCluster, isLeader),
                           Password = source.Password,
                           SkipIndexReplication = source.SkipIndexReplication,
                           TransitiveReplicationBehavior = source.TransitiveReplicationBehavior,
                           Url = source.Url,
                           Username = source.Username,
                           SpecifiedCollections = source.SpecifiedCollections,
                           ReplicateAttachmentsInEtl = source.ReplicateAttachmentsInEtl
                       };
            }
        }
    }

    /// <summary>
    /// Options for how to replicate replicated documents
    /// </summary>
    public enum TransitiveReplicationOptions
    {
        /// <summary>
        /// Don't replicate replicated documents
        /// </summary>
        [Description("Changed only")]
        None,
        /// <summary>
        /// Replicate replicated documents
        /// </summary>
        [Description("Changed and replicated")]
        Replicate
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/replication/ReplicationDestination.java
package net.ravendb.abstractions.replication;

import net.ravendb.abstractions.basic.UseSharpEnum;

import org.apache.commons.lang.builder.HashCodeBuilder;

public class ReplicationDestination {

  private String url;
  private String username;
  private String password;
  private String domain;
  private String apiKey;
  private String database;
  private TransitiveReplicationOptions transitiveReplicationBehavior;
  private Boolean ignoredClient;
  private Boolean disabled;
  private String clientVisibleUrl;

  public String getApiKey() {
    return apiKey;
  }

  public String getClientVisibleUrl() {
    return clientVisibleUrl;
  }

  public String getDatabase() {
    return database;
  }

  public Boolean getDisabled() {
    return disabled;
  }

  public String getDomain() {
    return domain;
  }

  public Boolean getIgnoredClient() {
    return ignoredClient;
  }

  public String getPassword() {
    return password;
  }

  public TransitiveReplicationOptions getTransitiveReplicationBehavior() {
    return transitiveReplicationBehavior;
  }

  public String getUrl() {
    return url;
  }

  public String getUsername() {
    return username;
  }

  public void setApiKey(String apiKey) {
    this.apiKey = apiKey;
  }

  public void setClientVisibleUrl(String clientVisibleUrl) {
    this.clientVisibleUrl = clientVisibleUrl;
  }

  public void setDatabase(String database) {
    this.database = database;
  }

  public void setDisabled(Boolean disabled) {
    this.disabled = disabled;
  }

  public void setDomain(String domain) {
    this.domain = domain;
  }

  public void setIgnoredClient(Boolean ignoredClient) {
    this.ignoredClient = ignoredClient;
  }

  public void setPassword(String password) {
    this.password = password;
  }

  public void setTransitiveReplicationBehavior(TransitiveReplicationOptions transitiveReplicationBehavior) {
    this.transitiveReplicationBehavior = transitiveReplicationBehavior;
  }

  public void setUrl(String url) {
    this.url = url.endsWith("/") ? url.substring(0, url.length() - 1) : url;
  }

  public void setUsername(String username) {
    this.username = username;
  }

  @UseSharpEnum
  public enum TransitiveReplicationOptions {

    // Don't replicate replicated documents
    NONE("Changed only"),

    // Replicate replicated documents
    REPLICATE("Changed and replicated");

    private String description;

    private TransitiveReplicationOptions(String description) {
      this.description = description;
    }

    public String getDescription() {
      return description;
    }

  }

  @Override
  public int hashCode() {
    return new HashCodeBuilder().append(apiKey).append(clientVisibleUrl)
      .append(database).append(disabled).append(domain)
      .append(ignoredClient).append(ignoredClient).append(password).append(transitiveReplicationBehavior).append(username).hashCode();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null) return false;
    if (getClass() != obj.getClass()) return false;
    ReplicationDestination other = (ReplicationDestination) obj;
    if (apiKey == null) {
      if (other.apiKey != null) return false;
    } else if (!apiKey.equals(other.apiKey)) return false;
    if (clientVisibleUrl == null) {
      if (other.clientVisibleUrl != null) return false;
    } else if (!clientVisibleUrl.equals(other.clientVisibleUrl)) return false;
    if (database == null) {
      if (other.database != null) return false;
    } else if (!database.equals(other.database)) return false;
    if (disabled == null) {
      if (other.disabled != null) return false;
    } else if (!disabled.equals(other.disabled)) return false;
    if (domain == null) {
      if (other.domain != null) return false;
    } else if (!domain.equals(other.domain)) return false;
    if (ignoredClient == null) {
      if (other.ignoredClient != null) return false;
    } else if (!ignoredClient.equals(other.ignoredClient)) return false;
    if (password == null) {
      if (other.password != null) return false;
    } else if (!password.equals(other.password)) return false;
    if (transitiveReplicationBehavior != other.transitiveReplicationBehavior) return false;
    if (username == null) {
      if (other.username != null) return false;
    } else if (!username.equals(other.username)) return false;
    return true;
  }



}
